habitat_baselines/common/obs_transformers.py

```diff
@@ -217,113 +218,545 @@ def from_config(cls, config: Config):
 
+        # Depth conversion
+        if ch == 1 and self.input_zfactors is not None:
+            self.input_zfactors = self.input_zfactors.to(batch.device)
+            batch = batch * self.input_zfactors.repeat(
+                batch_size // input_len, 1, 1, 1
+            )
+
         # Cache the repeated grids for subsequent batches
         if (
             self._grids_cache is None
             or self._grids_cache.size()[0] != batch_size
         ):
-            self._grids_cache = self.grids.repeat(batch_size // 6, 1, 1, 1)
+            self._grids_cache = self.grids.repeat(
+                batch_size // input_len, 1, 1, 1
+            )
             assert self._grids_cache.size()[0] == batch_size
         self._grids_cache = self._grids_cache.to(batch.device)
-        return self._to_equirec(batch)
+
+        out = self._convert(batch)
+        # Depth conversion
+        if ch == 1 and self.output_zfactors is not None:
+            output_b = batch_size // input_len
+            self.output_zfactors = self.output_zfactors.to(batch.device)
+            out = out * self.output_zfactors.repeat(output_b, 1, 1, 1)
+        return out
+
+    def calculate_zfactor(self, projections: List[CameraProjection]):
+        # z factor to convert depth in z value to depth from optical center
+        z_factors = []
+        for cam in projections:
+            if cam.depth_from == _DepthFrom.Z_VAL:
+                pts_on_sphere, _ = cam.unprojection(with_rotation=False)
+                zval_to_optcenter = 1 / pts_on_sphere[..., 2]
+                z_factors.append(zval_to_optcenter.unsqueeze(0))
+            else:
+                all_one = torch.full(
+                    (1, cam.img_h, cam.img_w), 1.0, dtype=torch.float
+                )
+                z_factors.append(all_one)
+        z_factors = torch.stack(z_factors)
+
+        if (z_factors == 1.0).all():
+            # All input cameras have depth from optical center
+            return None
+        else:
+            return z_factors
 
     def forward(self, batch: torch.Tensor) -> torch.Tensor:
-        return self.to_equirec_tensor(batch)
+        return self.to_converted_tensor(batch)
 
-    # Get theta and phi map
-    def get_theta_phi_map(self, equ_h: int, equ_w: int) -> torch.Tensor:
-        phi, theta = torch.meshgrid(torch.arange(equ_h), torch.arange(equ_w))
-        theta_map = -(theta + 0.5) * 2 * np.pi / equ_w + np.pi
-        phi_map = -(phi + 0.5) * np.pi / equ_h + np.pi / 2
-        return theta_map, phi_map
 
-    # Project on unit sphere
-    def angle2sphere(
-        self, theta_map: torch.Tensor, phi_map: torch.Tensor
-    ) -> torch.Tensor:
-        sin_theta = torch.sin(theta_map)
-        cos_theta = torch.cos(theta_map)
-        sin_phi = torch.sin(phi_map)
-        cos_phi = torch.cos(phi_map)
-        return torch.stack(
-            [cos_phi * sin_theta, sin_phi, cos_phi * cos_theta], dim=-1
+class _RotationMat(object):
```

Comment:
Maybe add a constructor / static factory method for generating these Tensors from arbitrary Euler angles / Radians as well>

Reply:
In which case, I guess it makes sense to have it be its own class instead of enum.

Reply:
I started to think that maybe we don't need `RotationMat`.  I agree with you that it's convenient to have methods to rotate matrix from arbitrary Euler angles / Radians, but [Rotaion in scipy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.transform.Rotation.html) should be enough. 

So, maybe only `get_cubemap_projections` like below is enough? (I couldn't think of the situation where only some rotation matrices out of six rotation matrices are used)
```python
def get_cubemap_projections():
    projections = []
    rotations = [
        torch.tensor([[-1, 0, 0], [0, 1, 0], [0, 0, -1]]),
        torch.tensor([[1, 0, 0], [0, 0, 1], [0, -1, 0]]),
        torch.tensor([[1, 0, 0], [0, 1, 0], [0, 0, 1]]),
        torch.tensor([[0, 0, -1], [0, 1, 0], [1, 0, 0]]),
        torch.tensor([[0, 0, 1], [0, 1, 0], [-1, 0, 0]]),
        torch.tensor([[1, 0, 0], [0, 0, -1], [0, 1, 0]]),
    ]
    for rot in rotations:
        cam = PerspectiveProjection(256, 256, R=rot)
        projections.append(cam)
    return projections
```

Reply:
Yeah, but definitely add some comments about which correspond to which direction and that is should BDFLRU ordering for the cubemap.

Reply:
Also maybe add a docstring / comment that these can be generated from scipy ordering (maybe even with a small rst code snippit).

Reply:
I added scipy code for generating rotation matrices. I hope this is fine.